### ISSUE TITLE
TCP conntrack does not honor window sizes correctly.

### ISSUE THREAD

**Issue by hbhasker** [google/gvisor#6734]
### Description

tcpconntrack package uses the segment's window size specified in the TCP segment header to determine if an incoming segment is acceptable. But it doesn't take into account any window scaling that might be in use as it completely ignores TCP options in the SYN(see: https://github.com/google/gvisor/blob/81f284f9d4af4170c86ad182e2f6f2cc9b3c565a/pkg/tcpip/transport/tcpconntrack/tcp_conntrack.go#L70).

Which means the check here https://github.com/google/gvisor/blob/81f284f9d4af4170c86ad182e2f6f2cc9b3c565a/pkg/tcpip/transport/tcpconntrack/tcp_conntrack.go#L227 will incorrectly reject packets as unacceptable for anything that might lie beyond the scaled down value of the *actual window*.

It also means this check https://github.com/google/gvisor/blob/81f284f9d4af4170c86ad182e2f6f2cc9b3c565a/pkg/tcpip/transport/tcpconntrack/tcp_conntrack.go#L227 will incorrectly determine the end of the acceptable window resulting in the FIN sequence being incorrect here

https://github.com/google/gvisor/blob/81f284f9d4af4170c86ad182e2f6f2cc9b3c565a/pkg/tcpip/transport/tcpconntrack/tcp_conntrack.go#L266

Which also means that the s.closed() will incorrectly return true once the sequence number that *we incorrectly assumed to be the fin sequence number* is ACKed. 

TLDR; we need to parse options in the SYN/SYN-ACK to determine if the window scaling is in effect and apply appropriate window scaling in each direction as its not required to be symmetric when calculating the actual window size.

**Comment by kevinGC** [comment]
Looks like we actually don't reject packets for being larger than the unscaled window. Because `tcpconntrack` was written pre-`PacketBuffer`, it assumes that passed-in `header.TCP` buffers also contain the payload. You can see this assumption when calculating the payload size:

https://github.com/google/gvisor/blob/81f284f9d4af4170c86ad182e2f6f2cc9b3c565a/pkg/tcpip/transport/tcpconntrack/tcp_conntrack.go#L329-L331

But we always pass only the TCP header without payload. The real bug is that we're always checking the window size against an apparently payload-less packet, so we never reject packets based on rwnd.

**Comment by kevinGC** [comment]
Fixed by 5e984d5aa2c3f98934040e199d460c74744e5799.
